### PR TITLE
[solax] Cloud connection support

### PR DESCRIPTION
**Pull request by theater**
Besides the direct connection to the local wi-fi module that we started to support with OpenHAB 4.1, now the binding will support also a new type - cloud connect. The documentation for the API can be obtained from the Solax web portal even though it's not very thorough.

========

Implemented a new thing type - cloud connect inverter.

Refactored the packages, model and classes to reflect that now we have two different ways to retrieve the data.

The common logic is extracted in the AbstractSolaxHandler.
The handler now requires two method implementations with the specific logics for the Local and Cloud connection:
* createConnector() - specific for the two different types of HTTP connection. 
*  updateFromData() - method that maps from the response to the specific thing type channels.

README is updated with the new type and the examples are up to date. Made a small corrections to other parts in the README.

I'm not sure if the thing type is new if I need to create an update instructions. I would expect to NOT need it.

### CODE REVIEW

**Comment by theater** [comment]
Depends on #16248 and has to be rebased on it once it's merged into main

**Comment by theater** [comment]
I have rebased everything on top of the change from the other PR which resulted in a very unpleasant fix of conflicts.
Now the PR is ready for a review. 

Hope someone finds time to review it before any other changes in main branch happen, so I don't have to do this again. It's very time consuming due to a lot of refactoring (including a package refactoring) in this PR.

Best regards,
Konstantin

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+			</parameter>
+			<parameter name="password" type="text" required="true">
+				<label>Password</label>
+				<description>Password for accessing cloud API (the serial number of the Wi-Fi module)</description>
+				<context>password</context>
+			</parameter>
+			<parameter name="token" type="text" required="true">
+				<label>Token</label>
+				<description>Token to access the Solax cloud API</description>
+				<context>network-address</context>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml:111]:
```suggestion
				<context>password</context>
```

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+				<label>Inverter Workmode</label>
+				<description>Inverter Workmode</description>
+			</channel>
+
+			<channel id="raw-data" typeId="raw-data-type"/>
+		</channels>
+
+		<properties>
+			<property name="thingTypeVersion">1</property>
+		</properties>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml:95]: Don't think this is needed as this is the first version if this type structure and no thing upgrade instructions are written for that first version.

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml [hunk 3]
```diff
@@ -54,6 +54,20 @@
 				<option value="11">PV Wake-up Battery</option>
 				<option value="12">GEN Check</option>
 				<option value="13">GEN Run</option>
+				<option value="100">Waiting</option>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml:57]: This channel now combines both local and cloud values into one channel-type / list . 

I think there should be two variants of this channel-type, one for cloud, one for local, each holder their own set of the values.

File: bundles/org.openhab.binding.solax/src/test/java/org/openhab/binding/solax/internal/cloud/TestCloudParser.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        assertEquals(151, bean.getInverterOutputPower(), "AC/Inverter output power");
+        assertEquals(0.9, bean.getYieldToday(), "Yield today");
+        assertEquals(7339.5, bean.getYieldTotal(), "Yield total");
+        assertEquals(-925.0, bean.getFeedInPower(), "Feed-in power");
+        assertEquals(147.2, bean.getFeedInEnergy(), "Feed-in energy");
+        assertEquals(2536.4, bean.getConsumeEnergy(), "Consume energy");
+        assertEquals(0, bean.getFeedInPowerM2(), "Feed in power M2");
+        assertEquals(56.3, bean.getEPSPowerR(), "EPS power R");
+        // assertEquals(ZonedDateTime.of(LocalDateTime.of(2023, 11, 28, 18, 34, 17), ZoneId.of("CET")),
+        // bean.getUploadTime(), "Upload time");
```
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/test/java/org/openhab/binding/solax/internal/cloud/TestCloudParser.java:87]: Can these tests be fixed?

**Review by lsiepel — CHANGES_REQUESTED** [review]
Did a full review. Most are minor comments. I also think the properties file lacks some values, so it is good to clean/re-run i18n plugin to be sure all are added correctly.

As this is on top of the previous PR and some conflicts have been resolved, i think it would be wise to spend some extra time on testing as we want to prevent regression.

File: bundles/org.openhab.binding.solax/README.md [hunk 5]
```diff
@@ -111,6 +111,40 @@ If you're missing a channel this means that it's not supported for your inverter
 | serialNumber      | The serial number of the Wi-Fi module     |
 | inverterType      | Inverter Type (for example X1_HYBRID_G4)  |
 
+
+### Cloud Connect Inverter Configuration
+
+| Parameter         | Description                                                                                                                                        |
+|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
+| refreshInterval   | Defines the refresh interval when the binding polls from from the Solax cloud (in seconds). Optional parameter(min=15, max=600). Default is 30 seconds. Be advised that the cloud API is limited to max 10 calls per minute and 10000 calls per day.                                                                                                              |
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/README.md:119]: It's to be on the safe side so we don't reach it due to OH restarts or refresh commands (not very likely though)... I guess 9 or maybe 10 (because it's more natural value) can be used....

File: bundles/org.openhab.binding.solax/README.md [hunk 6]
```diff
@@ -66,13 +66,13 @@ If you're missing a channel this means that it's not supported for your inverter
...
+| Channel                   | Type                       | Description                                                                                        |
+|---------------------------|----------------------------|----------------------------------------------------------------------------------------------------|
+| battery-power             | Number:Power               | The power to / from battery (negative means power is pulled from the battery and vice-versa) [W]   |
+| battery-current           | Number:ElectricCurrent     | The current to / from battery (negative means power is pulled from the battery and vice-versa) [A] |
+| battery-voltage           | Number:ElectricPotential   | The voltage of the battery [V]                                                                     |
+| battery-temperature       | Number:Temperature         | The temperature of the battery [C/F]                                                               |
+| battery-level             | Number                     | The state of charge of the battery [%]                                                             |
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/README.md:75]: > Is it worth the effort to introduce this beaking change to only rename the channel ?

It's not a breaking change but rather a fix in the documentation. Initially it was state of charge but has been changed. Wasn't change in the README though...

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public CloudHttpConnector(String uri, String tokenId, String serialNumber) {
+        this.uri = uri.replace(TOKEN_ID_REPLACEMENT_PATTERN, tokenId)
+                .replace(SERIAL_NUMBER_REPLACEMENT_PATTERN, serialNumber).trim();
+    }
+
+    @Override
+    public @Nullable String retrieveData() throws IOException {
+        logger.trace("Uri: {}", uri);
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:59]: Why we need this change? I prefer to keep the logic-related traces in DEBUG severity and the full request/response logs in TRACE severity.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    @Override
+    public @Nullable String getRawData() {
+        return rawData;
+    }
+
+    public boolean isError() {
+        String exception = getException();
+        return exception == null || exception.equals(ERROR);
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java:241]: It's a little bit stupid... they put in the exception field the value "success" if all is good and "error" if there is an error. Sorry if it's misleading but I would say that the API response is a bit stupid... Thing is that sometimes the response may be empty and I want to be sure to check that the exception is != null (therefore !error)

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        } else {
+            logger.debug("Unable to retrieve data because a request is already in progress.");
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            scheduler.execute(this::retrieveData);
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:113]: What is the idea here? To bring it offline forcefully with a proper status, so we don't end up in the API reject?

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml (hunk 3 above)
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml:57]: It's not really an issue because essentially they're the same but I will make it a separate thing type.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+ * sent to one of the channels.
+ *
+ * @author Konstantin Polihronov - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractSolaxHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractSolaxHandler.class);
+
+    private static final int INITIAL_SCHEDULE_DELAY_SECONDS = 5;
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:50]: hm... not really. I initially implemented it like this and it's a part of the extract from the old handler but can be immediate. I'll change it.

[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:50]: Btw I haven't seen any overloaded method without an initial delay. Does that mean I need to set the delay to 0?

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:59]: Just saw that actually in the Abstract handler it's a debug... will make this one a debug too....

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+public class CloudHttpConnector implements SolaxConnector {
+
+    private static final int HTTP_REQUEST_TIME_OUT = 5000;
+
+    // private static final String CONTENT_TYPE = "text/html; charset=utf-8";
+    private static final String CONTENT_TYPE = "application/json; charset=utf-8";
+
+    private final Logger logger = LoggerFactory.getLogger(CloudHttpConnector.class);
+
+    private static final String TOKEN_ID_REPLACEMENT_PATTERN = "{tokenId}";
```
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:39]: Hope I understood you correctly. I have removed the constants and put the string values directly in the "replace" methods in the constructor...

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java:241]: OK I refactored it a bit and also added a to the unit test some negative scenarios. I hope now it's more clear and robust. The issue is that I remember that there was a situation when the schedule was ending with an exception... hope I catched it now with the various unit tests...

**Comment by theater** [comment]
Hi @lsiepel, thanks very much for the fast review! 

I have fixed most of the comments, for some I left my own comments for clarification.

If you decide to meanwhile merge it, please hold until I do some integration testing of the changes, related to the exception management, so I don't pass any missed runtime exception... maybe I should just catch any exception in the AbstractSolaxHandler.retrieveData() method so I ensure another schedule attempt afterwards...

Cheers,
Konstantin

File: bundles/org.openhab.binding.solax/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/README.md:120]: In that case i would suggest to set the min to 9 as that is the absolute minimum. And set the default to 30 or so.

File: bundles/org.openhab.binding.solax/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/README.md:76]: Thanks!

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java]: Yes, set it to 0 = immediate.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:113]: Maybe if it is very close to the limit, bring ithe Thing offline until it is below some treshold (prevent flapping). That would mean you have to keep tracking of call timestamp. With 10k entry's that is maybe a but to expensive (performance wise).
It might get complex very fast. It's something to think about. The current solution is good enough to merge.

**Comment by theater** [comment]
Hi. I changed the requested minimum refresh interval to 9s, even though from practical perspective it doesn't make sense to refresh lower than 30-60 seconds, because the cloud API data gets updated from the local module once on each 5 minutes. :)
I checked the connectivity and for a whole night no issues (before when there was an issue it stopped working after an hour or so max), so I would expect also that the exception handling is OK and we're OK to merge now, unless you see anything else to add.

Thanks a lot for the review!

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM
